safedelete/queryset.py

```diff
@@ -27,13 +81,42 @@ def delete(self, force_policy=None):
             :py:func:`safedelete.models.SafeDeleteModel.delete`
         """
         assert self.query.can_filter(), "Cannot use 'limit' or 'offset' with delete."
-        # TODO: Replace this by bulk update if we can
-        for obj in self.all():
-            obj.delete(force_policy=force_policy)
-        self._result_cache = None
+
+        all_objs = self.all()
+        all_objs_id_list = list(all_objs.values_list('id', flat=True))
```

Comment:
Put it in the first if before it is used to avoid doing the query for every case.

Reply:
And because it is used in a queryset, you should not use list on it. It will make a subquery instead of a list.